lib/widgets/compose_box.dart

```diff
@@ -234,61 +307,110 @@ class _AttachFileButton extends StatelessWidget {
+      return;
+    }
+    if (result == null) {
+      return; // User cancelled; do nothing
     }
 
-    if (tooLargeFiles.isNotEmpty) {
-      final listMessage = tooLargeFiles
-        .map((file) => '${file.name}: ${(file.size / (1 << 20)).toStringAsFixed(1)} MiB')
-        .join('\n');
-      showErrorDialog( // TODO(i18n)
-        context: context,
-        title: 'File(s) too large',
-        message:
-          '${tooLargeFiles.length} file(s) are larger than the server\'s limit of ${store.maxFileUploadSizeMib} MiB and will not be uploaded:\n\n$listMessage');
+    if (context.mounted) {} // https://github.com/dart-lang/linter/issues/4007
+    else {
+      return;
     }
 
-    final List<(int, PlatformFile)> uploadsInProgress = [];
-    for (final file in rightSizeFiles) {
-      final tag = contentController.registerUploadStart(file.name);
-      uploadsInProgress.add((tag, file));
+    final Iterable<_File> files = result.files.map((f) {
+      assert(f.readStream != null);  // We passed `withReadStream: true` to pickFiles.
+      return _File(content: f.readStream!, length: f.size, filename: f.name);
+    });
+    _uploadFiles(context: context, contentController: contentController, contentFocusNode: contentFocusNode,
+        files: files);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return IconButton(icon: const Icon(Icons.image), onPressed: () => _handlePress(context));
+  }
+}
+
+class _AttachFromCameraButton extends StatelessWidget {
+  _AttachFromCameraButton({required this.contentController, required this.contentFocusNode});
+
+  final _picker = ImagePicker();
+  final ContentTextEditingController contentController;
+  final FocusNode contentFocusNode;
+
+  _handlePress(BuildContext context) async {
+    final XFile? result;
+    try {
+      // Ideally we'd open a platform interface that lets you choose between
+      // taking a photo and a video. `image_picker` doesn't yet have that
+      // option: https://github.com/flutter/flutter/issues/89159
+      // so just stick with images for now. We could add another button for
+      // videos, but we don't want too many buttons.
+      result = await _picker.pickImage(source: ImageSource.camera, requestFullMetadata: false);
+    } catch (e) {
+      if (e is PlatformException && e.code == 'camera_access_denied') {
+        // iOS has a quirk where it will only request the native
+        // permission-request alert once, the first time the app wants to
+        // use a protected resource. After that, the only way the user can
+        // grant it is in Settings.
+        showSuggestedActionDialog(context: context, // TODO(i18n)
+          title: 'Permissions needed',
+          message: 'To upload an image, please grant Zulip additional permissions in Settings.',
+          actionButtonText: 'Open settings',
+          onActionButtonPress: () {
+            AppSettings.openAppSettings();
```

Comment:
Looks like this is from a third-party library: https://pub.dev/packages/app_settings

Have you tried this out on both iOS and Android and it seems to work?

If at any point we're dissatisfied with how it works, the implementation is pretty simple — especially on iOS, where all the `AppSettings.openFoo` methods do the same thing:
https://github.com/spencerccf/app_settings/blob/ed65816d72933c9110dff9f4a74ee550e2af20b6/ios/Classes/SwiftAppSettingsPlugin.swift#L7-L10
but the Android version isn't complicated either:
https://github.com/spencerccf/app_settings/blob/ed65816d72933c9110dff9f4a74ee550e2af20b6/android/src/main/kotlin/com/example/appsettings/AppSettingsPlugin.kt#L45-L51

Reply:
`AppSettings.openAppSettings()` worked great on my iPhone 13 Pro running iOS 16.

Android office device (Samsung Galaxy S9, Android 9) seemed OK, but I wonder: have you ever had an app link directly into the screen for managing an app's permissions? This one brought me here:

<img width=400 src="https://user-images.githubusercontent.com/22248748/233753897-d7986dea-b444-4ffa-be81-0c0af07b5ab5.png">

and then I scrolled down and found a link to permissions:

<img width=400 src="https://user-images.githubusercontent.com/22248748/233753928-1e27393f-1d11-4018-849f-fefed07845fa.png">

which I tapped and found what I needed:

<img width=400 src="https://user-images.githubusercontent.com/22248748/233753940-7aebafdc-94f5-404b-b0aa-7f2ca13dcc86.png">

Reply:
> This one brought me here:

Same result when I try that function on my Pixel 5 running Android 13. Except that the subtitle on "Permissions" is "No permissions requested", because we're successfully using the no-privileges-needed API for this on newer Android versions. (To even call this function, I hacked up the code a bit.)


> I wonder: have you ever had an app link directly into the screen for managing an app's permissions?

Not sure. It's pretty uncommon for me to encounter an app linking me to any part of its system settings.

(Which doesn't mean we shouldn't do it; for one thing, this is an uncommon case, so it's quite possible that lots of apps offer the same thing if you do get into this case.)

Looking at docs, here's the intent type that the library is using for us:
https://developer.android.com/reference/android/provider/Settings#ACTION_APPLICATION_DETAILS_SETTINGS

Browsing around that page, here's a related intent type:
https://developer.android.com/reference/android/provider/Settings#ACTION_STORAGE_VOLUME_ACCESS_SETTINGS
which was deprecated, saying that "to manage storage permissions for a specific application" we should instead use… ACTION_APPLICATION_DETAILS_SETTINGS, just as `app_settings` is using for us.

There are various other intent types for particular permissions, including some storage-related:
https://developer.android.com/reference/android/provider/Settings#ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION
https://developer.android.com/reference/android/provider/Settings#ACTION_REQUEST_MANAGE_MEDIA

I don't think either of those correspond to what we'd want here, though. In particular both of those were introduced only in recent API versions, and I think they correspond to the newer more-explicitly-powerful permissions that were introduced at the same time as making normal applications stop needing such permissions.

Reply:
That reasoning makes sense.